### DISCUSSION TITLE
How to create multiple relationships to the same field

### DISCUSSION THREAD

**Discussion by shiminjia** [prisma/prisma#11108] [Q&A] [answered]
```prisma
model Like {
  id        String   @id @default(cuid())
  userId    String
  bookId    String  
  ...
  book      Book?     @relation(fields: [bookId], references: [id], name: "book") 
  comment   Comment?  @relation(fields: [bookId], references: [id], name: "comment") 

  @@index([slug])
}
```
please notice this. there are two relationships about `bookId`
```prisma
  book      Book?     @relation(fields: [bookId], references: [id], name: "book") 
  comment   Comment?  @relation(fields: [bookId], references: [id], name: "comment") 
```

ERROR：Error parsing attribute "@relation": The given constraint name `Like_bookId_fkey` has to be unique in the following namespace: global for foreign keys. Please provide a different name using the `map` argument.

Is it possible to create multiple relationships  to the same field `bookId` ?

**Comment by austincrim** [comment, marked as answer]
Hey @shiminjia 👋 ,

Since these two relationships are referencing the same exact fields, Prisma would generate a duplicate field name in the database.

You can fix the error by providing a `map` argument to one or both of the `@relation` directives in order to disambiguate the fields:
```prisma
model Like {
  id        String   @id @default(cuid())
  userId    String
  bookId    String  

  book      Book?     @relation(fields: [bookId], references: [id], name: "book", map: "like_book_fk") 
  comment   Comment?  @relation(fields: [bookId], references: [id], name: "comment", map: "like_comment_fk") 

  @@index([slug])
```

Let me know if you have questions!

**Reply by shiminjia** [reply]
Thank you so much

**Reply by JakubKoralewski** [reply]
@shiminjia were you able to actually create a `Like` and insert into the db? In mysql I'm getting a foreign key constraint P2003 error :(

**Reply by bencehusi** [reply]
@austincrim the solution give me a vague "foreign constraint failed" error.

Do you have any idea why? Here's what I try to acomplish:

```typescript
// Create a new list
const list = await prisma.list.create({
      data: {
        teamId,
    },
})

// Create a localization for the list
await prisma.listLocalization.create({
      data: {
        entityId: list.id,
        entityType: EntityType.LIST,
        languageCode: 'en',
        translation: name,
      },
})
```

The mapped fields are in the last two rows of `ListLocalization`.

```prisma
// The List model represents a collection of related ListItems. Each list belongs to a Team 
// and can be used to categorize Jobs using the JobCategory model.
model List {
  id           Int             @id @default(autoincrement())
  createdAt    DateTime        @default(now())
  updatedAt    DateTime        @updatedAt
  teamId       String
  team         Team            @relation(fields: [teamId], references: [id])
  items        ListItem[]
  localizations ListLocalization[] @relation("ListLocalizations")
  categories   JobCategory[]
}

// The ListItem model represents an individual item within a List. These items can be 
// assigned to a Job via the JobCategory model.
model ListItem {
  id             Int             @id @default(autoincrement())
  createdAt      DateTime        @default(now())
  updatedAt      DateTime        @updatedAt
  list           List            @relation(fields: [listId], references: [id], onDelete: Cascade)
  listId         Int
  localizations  ListLocalization[]  @relation("ListItemLocalizations")
  jobCategories  JobCategory[]
}

// The EntityType enum is used in the Localization model to specify whether a localization 
// refers to a List or a ListItem.
enum EntityType {
  LIST
  LISTITEM
}

// The ListLocalization model is used to store localized versions of Lists and ListItems. 
// The entityType field specifies whether the localization refers to a List or a ListItem.
model ListLocalization {
  id           Int        @id @default(autoincrement())
  entityId     Int
  entityType   EntityType
  languageCode String
  translation  String
  createdAt    DateTime   @default(now())
  updatedAt    DateTime   @updatedAt
  list         List?      @relation("ListLocalizations", fields: [entityId], references: [id], map: "ListLocalization_list_fkey")
  listItem     ListItem?  @relation("ListItemLocalizations", fields: [entityId], references: [id], map: "ListLocalization_listItem_fkey")
}
```

**Reply by ajoshi31** [reply]
@bencehusi  I am also getting the foreign key constraint of the other field If I pass another. If I pass list, I get error for listItem and if I pass listItem I get error for list.

@austincrim did you created ant triggered fucntion or its worked with prisma directly ?

error: 
foreign key constraint failed on the field: `group_sourceTypeId (index)` -----

Foreign key constraint failed on the field: `Category_sourceTypeId (index)` -----

**Reply by ajoshi31** [reply]
@shiminjia  did it actually worked ?

**Reply by igordelorenzi** [reply]
@ajoshi31, it probably didn't work, because the way it was pointed out, the same column (entityId) references two different tables (List and ListItem), causing a foreign key constraint conflict. For example, think that if you insert a reference 43 into the entityId column, this ID might exist in the List table but not necessarily in the ListItem table (that's what the foreign key constraint does, a match check for a future join).

**Reply by ilhamsa1** [reply]
I think it didn't work; I got an error: Foreign key constraint violated: balance_payout_fkey (index)

**Comment by nmfrankel** [comment]
With both `List`.`id`, `List`Item.`id` columns having an `Int` value, there will very likely be an overlap. So how do  you utilize the `entityType` in the foreign keys?

```prisma
model ListLocalization {
  id           Int        @id @default(autoincrement())
  ...
  entityId     Int
  entityType   EntityType
  ...
  list         List?      @relation("ListLocalizations", fields: [entityId], references: [id])
  listItem     ListItem?  @relation("ListItemLocalizations", fields: [entityId], references: [id])
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Comment by Abubakarraza** [comment]
Any updates on this? How can I fix the "Foreign key constraint violated" issue to allow multiple relationships using the same field?
